plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/PartitionMatchSupplier.java

```diff
@@ -0,0 +1,100 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.hive.util;
+
+import io.trino.plugin.hive.HiveColumnHandle;
+import io.trino.plugin.hive.HivePartition;
+import io.trino.spi.connector.DynamicFilter;
+
+import javax.annotation.Nullable;
+
+import java.util.List;
+import java.util.function.BooleanSupplier;
+
+import static io.trino.plugin.hive.HivePartitionManager.partitionMatches;
+import static java.util.Objects.requireNonNull;
+
+public final class PartitionMatchSupplier
+        implements BooleanSupplier
+{
+    private final DynamicFilter dynamicFilter;
+    private final HivePartition hivePartition;
+    private final List<HiveColumnHandle> partitionColumns;
+
+    @Nullable
+    private volatile Boolean finalResult; // value is null until the dynamic filter no longer needs to be evaluated
+
+    private PartitionMatchSupplier(DynamicFilter dynamicFilter, HivePartition hivePartition, List<HiveColumnHandle> partitionColumns)
+    {
+        this.dynamicFilter = requireNonNull(dynamicFilter, "dynamicFilter is null");
+        this.hivePartition = requireNonNull(hivePartition, "hivePartition is null");
+        this.partitionColumns = requireNonNull(partitionColumns, "partitionColumns is null");
+    }
+
+    @Override
+    public boolean getAsBoolean()
+    {
+        // Check for a final memoized result
+        Boolean result = this.finalResult;
+        if (result != null) {
+            return result;
+        }
+        return evaluateCurrentDynamicFilter();
+    }
+
+    private boolean evaluateCurrentDynamicFilter()
+    {
+        // Must be checked before calling dynamicFilter.getCurrentPredicate()
+        boolean filterIsComplete = dynamicFilter.isComplete();
+        boolean partitionMatches = partitionMatches(partitionColumns, dynamicFilter.getCurrentPredicate(), hivePartition);
+        if (!partitionMatches || filterIsComplete) {
+            // Store the result to avoid re-evaluating the filter for each subsequent split
+            this.finalResult = (Boolean) partitionMatches;
+        }
+        return partitionMatches;
+    }
+
+    public static BooleanSupplier createPartitionMatchSupplier(DynamicFilter dynamicFilter, HivePartition hivePartition, List<HiveColumnHandle> partitionColumns)
+    {
+        requireNonNull(dynamicFilter, "dynamicFilter is null");
+        requireNonNull(hivePartition, "hivePartition is null");
+        requireNonNull(partitionColumns, "partitionColumns is null");
+
+        if (partitionColumns.stream().noneMatch(dynamicFilter.getColumnsCovered()::contains)) {
+            return new ConstantBooleanSupplier(true);
+        }
+        if (dynamicFilter.isComplete()) {
+            // Evaluate the dynamic filter once and store the result as a constant
+            return new ConstantBooleanSupplier(partitionMatches(partitionColumns, dynamicFilter.getCurrentPredicate(), hivePartition));
```

Comment:
```
matches = partitionMatches(partitionColumns, dynamicFilter.getCurrentPredicate(), hivePartition);
return () -> matches;
```

Reply:
I avoided using lambdas for this and the case where no partition columns are referenced by the dynamic filter so that the returned `BooleanSupplier` interface will be at most, bi-morphic at usage sites instead of accidentally megamorphic as a result returning instances corresponding to two different lambdas in addition to `PartitionMatchSupplier`.

Reply:
is it worth code-commenting, eg adorn ConstantBooleanSupplier with a comment?

Reply:
Comment added

Reply:
BTW wouldn't it be a pretty low hanging fruit for the JVM to optimize `() -> const` lambdas?
Maybe the JVM already creates such a class under the covers?

cc @skrzypo987 @martint

Reply:
Since the generated class at the definition site must specifically implement whatever interface is required (in this case, `BooleanSupplier` but in other places, it could be `Supplier<Boolean>` or `Callable<Boolean>`, etc) you couldn't produce a single global `() -> const` lambda, but you could presumably produce one per target type. However, you also need to dedupe on the constant value itself (ie: `() -> true` and `() -> false`) which might be reasonable and beneficial for `BooleanSupplier` with boolean constants, but not for most other constant returning lambda types in general. I doubt that the JVM does this, for that reason- but I could be wrong.

That said, even if there were such an optimization in hotspot, we still have three potential implementations that we might produce:
- `() -> const`
- `() -> variable`
- `PartitionMatchSupplier`

Since hotspot will only devirtualize method dispatch at usage sites with 2 or fewer target methods, we don't really have a choice but to avoid `() -> const` in favor of the more general `() -> variable` form, and declaring the class manually makes that choice more explicit to the intent.